python/context_selector/context_widget.py

```diff
@@ -0,0 +1,956 @@
+        task_manager.task_failed.connect(self._on_task_failed)
+
+        # Query Shotgun for valid entity types for PublishedFile.entity field
+        self._schema_query_id = task_manager.add_task(
+            _query_publishedfile_entity_schema)
+
+        # query all my assigned tasks in a bg task
+        self._my_tasks_query_id = task_manager.add_task(_query_my_tasks)
+
+        # get recent contexts from user settings
+        self._get_recent_contexts()
+
+    @property
+    def context_label(self):
+        """
+        The label for the context widget.
+        """
+        return self.ui.label
+
+    def enable_editing(self, enabled, message=None):
+        """
+        Show/hide the input widgets and display a message in the context label.
+        """
+
+        if enabled:
+            self.ui.edit_widget.show()
+        else:
+            self.ui.edit_widget.hide()
+
+        self.context_label.setText(message or "")
+
+    def _add_to_recents(self, context):
+        """
+        Adds the supplied context as an action in the list of recents context
+        actions
+        """
+
+        # don't add a "project" context to recents. we shouldn't encourage it
+        if context.project and not context.entity and not context.task:
+            return
+
+        logger.debug("Adding context to 'Recents': %s" % (context,))
+
+        recent_actions = self._menu_actions["Recent"]
+
+        matching_indexes = []
+        for i, recent_action in enumerate(recent_actions):
+            recent_context = recent_action.data()
+            if recent_context == context:
+                # contexts support __eq__ so this should be enough for comparing
+                matching_indexes.append(i)
+
+        if matching_indexes:
+            # context exists in recent list in one or more places. remove the
+            # QAction(s) and put one of them at the front of the list
+            recent_action = None
+            for match_index in matching_indexes:
+                recent_action = recent_actions.pop(match_index)
+        else:
+            # the context does not exist in the recents. add it
+            recent_action = self._get_qaction_for_context(context)
+
+        if recent_action:
+            recent_actions.insert(0, recent_action)
+
+        # only keep the 5 most recent
+        self._menu_actions["Recent"] = recent_actions[:5]
+
+    def _build_actions(self, tasks, group_name, sort=False,
+        exclude_current_context=False):
+        """
+        Build a list of actions from the supplied tasks. The actions are stored
+        in the instance's _menu_actions dictionary and used to build the menu.
+
+        The actions will be sorted by name if ``sort`` is set to True. If the
+        ``exclude_current_context`` is supplied, the widget's current context
+        will not be included in the list of actions.
+        """
+
+        publisher = sgtk.platform.current_bundle()
```

Comment:
maybe we define `self._bundle` in `__init__()` instead? we should rename this at least.

Reply:
all renamed to `bundle`